### PR TITLE
restore + dependency lock + gc design spec

### PR DESCRIPTION
**Pull request by OsmondJiang**
#3888

### CODE REVIEW

**Comment by OsmondJiang** [comment]
for build service using dependency lock:

- for non-pr build, setup a no-existing locale `dependencyLock` file path in config and docfx `restore` will generate one in that place.

- for pr build, setup the `dependencyLock`  to a remote url or a locale path, use the `--no-save` to skip `updating` step.

File: docs/designs/restore-and-gc.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            "downloads":{
+                "https:///some-blob-service/build-history.json":{
+                    "version": "66458170ba048c7fc2de4+%18b265bc4e9a6b4fdff8b5210caf9e357%22"
+                }
+            }
+        }
+    },
+    "downloads":{
+        "https:///some-blob-service/github-user-cache.json":{
+            "version": "18b265bc4e9a6bf4cb2d+%22f02fde5b4f3817266458170ba048c7fc2de46287%22"
```
Comment (yufeih) [inline docs/designs/restore-and-gc.md]: the example here is not a valid hash string,

File: docs/designs/restore-and-gc.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        }
+    }
+}
+```
+
+- `git` is the version mappings of dependency repositories
+- `downloads` is the version mappings of dependency files
+- `version` is `commit id` of the dependency repository  or `content sha1` of the dependency file
```
Comment (yufeih) [inline docs/designs/restore-and-gc.md]: shall we name them `commit` and `sha1`/`hash` respectively then?

File: docs/designs/restore-and-gc.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+This file describes an `exact`, and more importantly `reproducible` `dependencies` trees. Once it's present, any future `restore` will base its work off this file, instead of recalculating`.
+
+The presence of the `dependency lock` changes the `restore` behavior:
+
+- The dependency files/repository described by this `dependency lock` is reproduced, which means reproducing the structure/version described in the file.
```
Comment (yufeih) [inline docs/designs/restore-and-gc.md]: add a section that lock file generation process `docfx.yml + .lock.json --> .lock.json` should be _idempotent_, and that content inside `.lock.json` is always be sorted.

File: docs/designs/restore-and-gc.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+You can setup your `.lock.json` file(local file or a URL) in the configuration like
+
+```yml
+dependencyLock: https://some-blob-service/repo/branch/.lock.json
+``` 
+
+or 
+
+```yml
+dependencyLock: .lock.json
```
Comment (yufeih) [inline docs/designs/restore-and-gc.md]: add that the default value for `dependencyLock` is _null_, and by default, `docfx restore` will not generate a lock file.

File: docs/designs/restore-and-gc.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+```
+
+Whenever you run docfx `restore`, docfx will:
+
+- [GET] **Try** to get the `.lock.json` file from configuration, if it's present, any future dependencies restore will base its work off this file.
+- [UPDATE] Once finished the restore, docfx will **try** to create/update the `.lock.json` file defined in configuration, no matter it's a local or remote location.
```
Comment (yufeih) [inline docs/designs/restore-and-gc.md]: `githubUserCache` has a flag `updateRemoteUserCache=false`  to determine whether to update remote cache. Maybe similar behavior here as well?

File: docs/designs/restore-and-gc.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+For the last step [UPDATE], you can use command `--no-save` to skip.
```
Comment (yufeih) [inline docs/designs/restore-and-gc.md]: `npm` generates lock files by default, so it has a `--no-save` shortcut to disable it. We are quite the opposite: we don't generate lock files by default, unless you set `dependencyLock` somewhere, so maybe we don't need this option.

File: docs/designs/restore-and-gc.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
+# Restore
+
+Docfx `restore` helps you to restore all you dependency repositories/files(defined in [config](ocs/designs/config.md), fallback source repository and theme repository) into your local `%DOCFX_APPDATA_PATH%` for build preparation.
```
Comment (superyyrrzz) [inline docs/designs/restore-and-gc.md]: `[config](ocs/designs/config.md)` invalid link

Comment (OsmondJiang) [reply docs/designs/restore-and-gc.md]: changed. 赞

---
In reply to: [247026326](https://github.com/dotnet/docfx/pull/3900#discussion_r247026326) [](ancestors = 247026326)

File: docs/designs/restore-and-gc.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            | - master-fd6968dec9a9a39aec1845232466fe35fca520da
+            | - live-a82cb2a51f6925b2fa87275218b01678c5aadfe1
+```
+
+- `{url-short-path}` is calculated from `git remote` url
+- `{worktree-name}` is combined by `branch name` and `HEAD commit id`
+
+## Dependency lock
+
+There are some cases you want to keep using the **same version** of dependency files/repository between different `build`s.
```
Comment (superyyrrzz) [inline docs/designs/restore-and-gc.md]: This works for git, but not for file. We can go to a specific commit for a git repo, but cannot ask server to return a file with a specific e-tag existed before.

So, lock file can only consider git. DocFX always fetches latest for file dependency if cache miss.

Comment (live1206) [reply docs/designs/restore-and-gc.md]: For files, could `Etag` work similar to `version`?

File: docs/designs/restore-and-gc.md (hunk 2 above)
Comment (OsmondJiang) [reply docs/designs/restore-and-gc.md]: good idea

---
In reply to: [247021515](https://github.com/dotnet/docfx/pull/3900#discussion_r247021515) [](ancestors = 247021515)

File: docs/designs/restore-and-gc.md (hunk 8 above)
Comment (OsmondJiang) [reply docs/designs/restore-and-gc.md]: already have NOTE about that

---
In reply to: [247030971](https://github.com/dotnet/docfx/pull/3900#discussion_r247030971) [](ancestors = 247030971)

File: docs/designs/restore-and-gc.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+```text
+`%DOCFX_APPDATA_PATH%`
+    | - downloads
+        | - raw.gith..tent.com+docascode+docfx-te..ndencies+62b0448+extend1.yml+dc363b0e
+            | - 18b265bc4e9a6b4fdff8b5210caf9e3541f4cb2d+%22f02fde5b4f3817266458170ba048c7fc2de46287%22
+            | - 3111eb522b092c9cef5cd0ccd86892cf847592d6+%2299a5dbb8d73da1ec452d74c0ab45b11ebe8615a0%22
+
+```
+
+- `{url-short-path}` is calculated from `file` url
```
Comment (928PJY) [inline docs/designs/restore-and-gc.md]: >is calculated from `file` url [](start = 21, length = 29)

what is the calculate logic? should we add it here?

Comment (OsmondJiang) [reply docs/designs/restore-and-gc.md]: the above `restore file` and `restore git`

**Pull request merged**